components/script/dom/htmlimageelement.rs

```diff
@@ -356,20 +390,244 @@ impl HTMLImageElement {
+                    source_set.image_sources.push(ImageSource {
+                        url: element.get_string_attribute(&local_name!("src")).to_string(),
+                        descriptor: Descriptor { wid: None, den: None }
+                    })
+                }
+
+                // Step 4.1.4
+                Self::normalise_source_densities(&mut source_set);
+
+                // Step 4.1.5
+                *self.source_set.borrow_mut() = source_set;
+
+                // Step 4.1.6
+                return ;
+            }
+            // Step 4.2
+            if !element.is::<HTMLSourceElement>() {
+                continue;
+            }
+
+            // Step 4.3 - 4.4
+            let mut source_set = HTMLImageElement::new_source_set();
+            match element.get_attribute(&ns!(), &local_name!("srcset")) {
+                Some(x) => {
+                    source_set.image_sources = parse_a_srcset_attribute(&x.value());
+                }
+                _ => continue
+            }
+
+            // Step 4.5
+            if source_set.image_sources.len() == 0 {
+                continue;
+            }
+
+            // Step 4.6
+            match element.get_attribute(&ns!(), &local_name!("media")) {
+                Some(x) => {
+                    // Check if value of the attribute matches the environment
+                    if self.matches_environment(x.value().to_string()) {
+                        ()
+                    } else {
+                        continue;
+                    }
+                }
+                _ => ()
+            }
+
+            // Step 4.7
+            match element.get_attribute(&ns!(), &local_name!("sizes")) {
+                Some(x) => {
+                    source_set.source_size =
+                        parse_a_sizes_attribute(DOMString::from_string(x.value().to_string()));
+                    self.set_value_source_size_list(& mut source_set.source_size);
+                }
+                _ => ()
+            }
+
+            // Step 4.8
+            match element.get_attribute(&ns!(), &local_name!("type")) {
+                Some(x) => {
+                    match get_mime_type_str(&x.value()) {
+                        Some(_valid_mime) => (),
+                        None => continue // Unsupported mime type
+                    }
+                }
+                _ => continue // Unknown mime type
+            }
+
+            // Step 4.9
+            Self::normalise_source_densities(&mut source_set);
+
+            // Step 4.10
+            *self.source_set.borrow_mut() = source_set;
+            return ;
+        }
+    }
+
+    fn set_value_source_size_list(&self, source_size_list: &mut SourceSizeList) {
+        let document = document_from_node(self);
+        let device = document.device().unwrap();
+        let quirks_mode = document.quirks_mode();
+        let length = source_size_list.evaluate(&device, quirks_mode);
```

Comment:
This function seems like it's a bit backwards right now. `evaluate` will return a final size that matches a particular media condition, or a fallback value (SourceSizeList::value) if it exists, or default to a value equivalent to 100vw otherwise. That means that we should accept the fallback value as an argument to this method, set the value field before calling `evaluate`, and then return the evaluated value from the function, rather than overwriting the value field.

Reply:
I am a little confused, where will the value returned by this evaluate method get used? Should we call this method in normalise_source_densities ?